packages/jest-cli/src/watch.js

```diff
@@ -33,41 +33,30 @@ const watch = (
   stdin?: stream$Readable | tty$ReadStream = process.stdin
 ) => {
   setWatchMode(argv, argv.watch ? 'watch' : 'watchAll', {
-    pattern: argv._,
+    testNamePattern: argv.testNamePattern,
+    testPathPattern: argv.testPathPattern
+      || (argv._ instanceof Array ? argv._.join('|') : ''),
   });
-  let currentPattern = '';
+
+  let isEnteringString = false;
```

Comment:
As pattern mode becomes more and more complex it seems that we need to start extracting all of this logic out of watch.js.. so that the pattern mode and typeahead stuff can scale and be maintained in the future.